src/shogun/mathematics/linalg/LinalgBackendViennacl.h

```diff
@@ -115,6 +186,27 @@ class LinalgBackendViennaCL : public LinalgBackendGPUBase
 		return viennacl::linalg::inner_prod(a_gpu->data(a.size()), b_gpu->data(b.size()));
 	}
 
+	/** Return the largest element in the vector with ViennaCL library */
+	template <typename T>
+	T max_impl(const SGVector<T>& vec) const
+	{
+		viennacl::ocl::kernel& kernel = generate_kernel<T>();
+
+		SGVector<T> result(1);
+		result.set_const(-1);
+		GPUMemoryViennaCL<T>* result_gpu =
+			static_cast<GPUMemoryViennaCL<T>*>(linalg::to_gpu(result).gpu_vector.get());
+		GPUMemoryViennaCL<T>* vector_gpu =
+			static_cast<GPUMemoryViennaCL<T>*>(vec.gpu_vector.get());
+
+		viennacl::ocl::enqueue(kernel(vector_gpu->data(vec.size()),
+			cl_int(vec.size()), cl_int(vector_gpu->m_offset),
+			result_gpu->data(1)));
+
+		SGVector<T> result_cpu = linalg::from_gpu(new SGVector<T>(result_gpu, 1));
+		return result_cpu[0];
+	}
```

Comment:
- `GPUMemoryViennaCL<T>` can only be accessed (initialized) from existed `SGVector<T>`, which makes the code complicated. To save the result, I need construct a CPU vector with fixed length and value, and then transfer it to the GPU. Things will be easier if there's variable in `GPUMemoryViennaCL<T>` to hold the vector length.

- There's problem with `SGVector<T>(GPUMemoryBase<T>* vector, index_t len)` constructor. As the cpu vector pointer is not allocated, it cannot be freed. (`malloc: *** error for object 0x7fe0c2c18320: pointer being freed was not allocated`). I still need to work on it.